appscripts/canalytics.gs

```diff
@@ -0,0 +1,312 @@
+          break;
+        }
+               
+      }
+
+    }
+  } // End of iterating over all data in sheet
+  // Must remember to replace the header!!
+  data.unshift(header)
+  sheet.getDataRange().setValues(data);
+  // TODO: Update OPRs?
+
+}
+
+function getOPRS() {
+  // TODO:  Get a new OPRS with timestamp everytime it changes?  So we can see the change of OPRS over time?
+  // WOuld need to check the cache, I guess?
+
+}
+
+function eventFinalMatches() {
+  // TODO: This is duplicating the work done in eventQualMatches.  Should try to refactor to not duplicate code
+  // 
+}
+
+// To get a scoreBreakdown to show up in the spreadsheet, both a header needs to be added, and
+// the value needs to be calcualted in the scoreBreakdown function
+function scoreBreakdownHeader(){
+  // return year specific score breakdown header
+  return ["red1_taxi","red2_taxi","red3_taxi",
+    "red1_endgame","red2_endgame","red3_endgame",
+    "blue1_taxi","blue2_taxi","blue3_taxi",
+    "blue1_endgame","blue2_endgame","blue3_endgame"]    
+}
+function scoreBreakdown(match){
+  // Breakdown year specific score information.  Set up for 2022 season
+  var sbd=match.score_breakdown
+  match.red1_taxi = sbd.red.taxiRobot1
+  match.red2_taxi = sbd.red.taxiRobot2
+  match.red3_taxi = sbd.red.taxiRobot3
+  match.blue1_taxi = sbd.blue.taxiRobot1
+  match.blue2_taxi = sbd.blue.taxiRobot2
+  match.blue3_taxi = sbd.blue.taxiRobot3
+  match.red1_endgame = sbd.red.endgameRobot1
+  match.red2_endgame = sbd.red.endgameRobot2
+  match.red3_endgame = sbd.red.endgameRobot3
+  match.blue1_endgame = sbd.blue.endgameRobot1
+  match.blue2_endgame = sbd.blue.endgameRobot2
+  match.blue3_endgame = sbd.blue.endgameRobot3
+
+}
+/**
+ * Handles the HTTPS request for TBA requests. Takes the path, such as team/frc1234/simple, and returns the resulting JSON object. 
+ * Notes about caching:
+ *   We cache every response with a key of the url
+ *   We save the current time + max-age from the Cache-Control header of the response in the Document Cache Service as well as the ETag
+ *   When we check again, if the time that has elapsed is less than the max-age, then we will return the cached data.
+ *   If the cache is no longer valid, we will check with The Blue Alliance, with the Etag from the last response.
+ *   ( reference: https://www.thebluealliance.com/apidocs look for Caching )
+ *   If the response code is 304, then the data has not changed, and the Cache is still valid even though the time has passed.
+ *   Since the data can be valid past the max-age, we lie to the cache-control and ask it to save data for 6 hours.
+ * 
+ * @param {text} path The part of the URL after "/api/v3/" in the query.
+ * @return Parsed json of the result including a cache time and Etag used for reducing data transfer.
+ * @customfunction
+ */
+function TBAQuery(path) {
+  var url = 'https://www.thebluealliance.com/api/v3/'+path
+  const cacheDocument = CacheService.getDocumentCache();
```

Comment:
Since cache problems will probably be fairly common, a Clear Cache function would probably be a good early addition

Reply:
I couldn't find a way to print out all keys that had been cached. 

I think that an ignoreCache flag in TBAQuery, and using that for the initial data pulls would be a good addition.  Will add it

Reply:
Added an ignoreCache Flag to many of the routines including TBAQuery